backends/guest/common/read.c

```diff
@@ -111,6 +111,77 @@ int print_cert_info(crypto_x509_t *x509)
 	return SUCCESS;
 }
 
+/*
+ * prints ESDs inside an ESL buffer in human readable form
+ *
+ * @param esl, pointer to the beginning of valid esl data. NOTE: caller is responsible for validation
+ * @param esl_size, size of the esl data buffer
+ * @param sig_type, signature type enum as extracted from e.g. get_signature_type
+ * @return SUCCESS or propogates error code
+ */
+static int print_esd_from_esl_buffer(const uint8_t *esl, size_t esl_size,
+				     enum signature_type sig_type)
+{
+	crypto_x509_t *x509 = NULL;
+	int rc;
+	size_t esd_data_size, esd_count = 0;
+	uuid_t esd_owner;
+	union {
+		const uint8_t *raw;
+		sv_esd_t *esd;
+	} curr_esd;
+	curr_esd.raw = NULL;
+
+	rc = next_esd_from_esl(esl, &curr_esd.raw, &esd_data_size, &esd_owner);
+	if (rc) {
+		prlog(PR_ERR, "Error reading esd from esl, rc = %d\n", rc);
+		return rc;
+	}
+
+	while (curr_esd.esd != NULL) {
+		esd_count++;
+		switch (sig_type) {
+		case ST_HASHES_START ... ST_HASHES_END:
+			printf("\tData-%zu: ", esd_count);
+			print_hex(curr_esd.raw, esd_data_size);
+		case ST_X509:
+			x509 = crypto_x509_parse_der(curr_esd.raw, esd_data_size);
+			if (!x509)
+				break;
+			printf("\tCertificate-%zu: ", esd_count);
+			rc = print_cert_info(x509);
+
+			// we're done with the x509, free it immediately
+			crypto_x509_free(x509);
+			x509 = NULL;
+
+			// ...then bail if there was an error printing the cert
+			if (rc)
+				return rc;
+
+			break;
+		case ST_SBAT:
+			printf("\tData: ");
+			print_raw((char *)curr_esd.raw, esd_data_size);
+			break;
+		case ST_DELETE:
+			printf("\tDELETE-MSG: ");
+			print_raw((char *)curr_esd.raw, esd_data_size);
+			break;
+		default:
+			prlog(PR_ERR, "ERROR: invalid signature type = %d\n", sig_type);
+			break;
+		}
+
+		rc = next_esd_from_esl(esl, &curr_esd.raw, &esd_data_size, &esd_owner);
+		if (rc) {
+			prlog(PR_ERR, "Error reading next esd (%zu), rc = %d\n", esd_count, rc);
+		}
```

Comment:
We should probably break/return here or change the `while` conditional to check `rc`.
The loop will go as long as `curr_esd` is not `NULL`, there are several cases where `next_esd_from_esl` can return an errorcode and not update the return pointer to be NULL

Reply:
Ah dangit, knew I missed one.

Also changing the error message for the default case nearby to be "unknown" signature type, since that's the one "error" we're allowing without exiting.